File: Linked Lists/876. Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# Using two passes

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        length = 0
        
        curr = head
        while curr:
            length += 1
            curr = curr.next
        
        mid = length // 2

        curr = head
        for _ in range(mid):
            curr = curr.next
        
        return curr

# Time: O(n)
# Space: O(1)


# Using Floyd's Cycle Finding Algorithm / Hare-Tortoise Algorithm:

class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        return slow

# Time: O(n)
# Space: O(1)

File: Linked Lists/876. Middle of the Linked List/876. Middle of the Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Floyd's Cycle Finding Algorithm / Hare-Tortoise Algorithm:
var middleNode = function (head) {
  let fast = head;
  let slow = head;

  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow.next;
  }

  return slow;
};

// Time: O(n)
// Space: O(1)
